airflow/operators/python.py

```diff
@@ -412,140 +427,113 @@ def __init__(  # pylint: disable=too-many-arguments
-                                   "different Python major versions "
-                                   "for PythonVirtualenvOperator. "
-                                   "Please use string_args.")
+        if not self.system_site_packages and self.use_dill and 'dill' not in self.requirements:
+            self.requirements.append('dill')
+        self.pickling_library = dill if self.use_dill else pickle
 
     def execute_callable(self):
         with TemporaryDirectory(prefix='venv') as tmp_dir:
             if self.templates_dict:
                 self.op_kwargs['templates_dict'] = self.templates_dict
-            # generate filenames
+
             input_filename = os.path.join(tmp_dir, 'script.in')
             output_filename = os.path.join(tmp_dir, 'script.out')
             string_args_filename = os.path.join(tmp_dir, 'string_args.txt')
             script_filename = os.path.join(tmp_dir, 'script.py')
 
-            # set up virtualenv
-            python_bin = 'python' + str(self.python_version) if self.python_version else None
             prepare_virtualenv(
                 venv_directory=tmp_dir,
-                python_bin=python_bin,
+                python_bin=f'python{self.python_version}' if self.python_version else None,
                 system_site_packages=self.system_site_packages,
-                requirements=self.requirements,
+                requirements=self.requirements
             )
 
             self._write_args(input_filename)
-            self._write_script(script_filename)
             self._write_string_args(string_args_filename)
+            self._write_script(script_filename)
+
+            execute_in_subprocess(cmd=[
+                f'{tmp_dir}/bin/python',
+                script_filename,
+                input_filename,
+                output_filename,
+                string_args_filename
+            ])
 
-            # execute command in virtualenv
-            execute_in_subprocess(
-                self._generate_python_cmd(tmp_dir,
-                                          script_filename,
-                                          input_filename,
-                                          output_filename,
-                                          string_args_filename))
             return self._read_result(output_filename)
 
-    def _pass_op_args(self):
-        # we should only pass op_args if any are given to us
-        return len(self.op_args) + len(self.op_kwargs) > 0
+    def _write_args(self, filename):
+        if self.op_args or self.op_kwargs:
+            if self.op_kwargs:
+                # some items from context can't be loaded in virtual env
+                self._keep_serializable_op_kwargs()
+            with open(filename, 'wb') as file:
+                self.pickling_library.dump({'args': self.op_args, 'kwargs': self.op_kwargs}, file)
+
+    def _keep_serializable_op_kwargs(self):
```

Comment:
Now we're removing all kinds of keys, wouldn't it be more preferable to have an allowed-list for keys to keep? This might be more stable in the future.

Reply:
Good point. 🤔

Reply:
But the problem is that you can add custom `op_kwargs` which should also be serialized but you don't know how they are called in the first place. 😁 
```python
    def execute(self, context: Dict):
        context.update(self.op_kwargs)
        ...
        self.op_kwargs = PythonOperator.determine_op_kwargs(self.python_callable, context, len(self.op_args))

        return_value = self.execute_callable()
        ...
```
The `PythonVirtualenvOperator` can only access `self.op_kwargs` not `context` directly which contains those unserializable objects.
I would have to overwrite the `execute` so that I can change that.

WDYT? @Fokko

Reply:
I think it is actually not that hard to do. :)